keps/sig-scheduling/5004-dra-extended-resource/README.md

```diff
@@ -0,0 +1,1061 @@
+resources backed by DRA on a node, i.e. those that are advertised by `ResourceSlice`.
+
+```go
+type Resource struct {
+	MilliCPU         int64
+	Memory           int64
+	EphemeralStorage int64
+	// We store allowedPodNumber (which is Node.Status.Allocatable.Pods().Value())
+	// explicitly as int, to avoid conversions and improve performance.
+	AllowedPodNumber int
+	// ScalarResources
+	ScalarResources map[v1.ResourceName]int64
+
+	// NEW!
+	// DynamicResources: keep track of extended resources backed by DRA to device class
+	// The map's key is the extended resource name that has exactly one device
+	// class advertises it.
+	DynamicResources map[v1.ResourceName]string
+}
+```
+
+type `NodeInfo` is used by scheduler to keep track of the information for each
+node in memory. Its `Allocatable` field is used to keep track of the allocatable
+resources in memory. At the beginning of each scheduling cycle, scheduler takes
+a snapshot of all the nodes in the cluster, and updates their corresponding
+`NodeInfo`.
+
+For the scheduler with DRA enabled, right after taking the node snapshot, the
+scheduler also takes a snapshot of `DeviceClass`, and updates
+`NodeInfo.DynamicResources` if there is an extended resource backed by DRA.
+
+For a node with extended resources from device plugin, its NodeInfo's
+Allocatable.ScalarResources is updated with the k8s `Node`'s object.
+For a node with extended resources backed by DRA, its NodeInfo's
+Allocatable.DynamicResources is updated based on DRA `DeviceClass` objects.
+
+The existing 'noderesources' plugin needs to be modified, such that a pod's
+extended resource request is checked against a NodeInfo's ScalarResources if the
+node uses device plugin, and checked against a NodeInfo's DynamicResources, if
+the request is for extended resources backed by DRA, then 'noderesources' plugin
+would pass and leave it to 'dynamicresource' plugin to check if it can be satisfied.
+
+The existing 'dynamicresources' plugin needs to be modified to account for the
+extended resource backed by DRA requests.
+
+#### EventsToRegister
+This registers all cluster events that might make an unschedulable pod schedulable,
+like finishing the allocation of a claim, or resource slice updates.
+
+The existing dynamicresource plugin has registered all the events needed or
+extended resource backed by DRA, hence no change is needed.
+
+#### PreFilter
+It checks if the pod has any container requests for extended resources backed by DRA.
+If not, and no claims in the pod, then the plugin can return early, as there
+is nothing to do.
+
+If the pod still needs to be considered by the plugin, then it checks if the
+special resource claim for extended resources backed by DRA has been created
+before by scheduler, by checking resource claim name having pod name in the
+annotation `resource.kubernetes.io/extended-resource-claim: pod-name`.
+
+If found, scheduler would reuse it. If not found, scheduler would create a
```

Comment:
Please clarify this "create" means just in-memory thingy on the KEP.
I was about to make the same question as @/liggitt here before I saw:
https://github.com/kubernetes/enhancements/pull/5136/files#r2091951508

Reply:
And, also how is this in-memory ResourceClaim implemented? Through the scheduler cache?